doc/mount.glusterfs.8
.\"  Copyright (c) 2008-2013 Red Hat, Inc. <http://www.redhat.com>
.\"  This file is part of GlusterFS.
.\"
.\"  This file is licensed to you under your choice of the GNU Lesser
.\"  General Public License, version 3 or any later version (LGPLv3 or
.\"  later), or the GNU General Public License, version 2 (GPLv2), in all
.\"  cases as published by the Free Software Foundation.
.\"
.\"
.\"
.TH GlusterFS 8 "Cluster Filesystem" "14 September 2013" "Red Hat, Inc."
.SH NAME
.B mount.glusterfs - script to mount native GlusterFS volume
.SH SYNOPSIS
.B mount -t glusterfs [-o <options>] <volumeserver>:/<volume>[/<subdir>]
.B         <mountpoint>
.TP
.B mount -t glusterfs [-o <options>] <server1>,<server2>,
.B        <server3>,..<serverN>:/<volname>[/<subdir>] <mount_point>
.TP
.TP
.B mount -t glusterfs [-o <options>] <path/to/volumefile> <mountpoint>
.PP
.SH DESCRIPTION
This tool is part of \fBglusterfs\fR(8) package, which is used to mount using
GlusterFS native binary.

\fBmount.glusterfs\fR is meant to be used by the mount(8) command for mounting
native GlusterFS client. This subcommand, however, can also be used as a
standalone command with limited functionality.

.SH OPTIONS
.PP
.SS "Basic options"
.PP
.TP
\fBlog\-file=\fRLOG-FILE
File to use for logging [default:/var/log/glusterfs/glusterfs.log]
.TP
\fBlog\-level=\fRLOG-LEVEL
Logging severity.  Valid options are TRACE, DEBUG, WARNING, ERROR, CRITICAL
INFO and NONE [default: INFO]
.TP
\fBacl
Mount the filesystem with POSIX ACL support
.TP
\fBfopen\-keep\-cache[=BOOL]
Do not purge the cache on file open (default: false)
.TP
\fBworm
Mount the filesystem in 'worm' mode
.TP
\fBaux\-gfid\-mount
Enable access to filesystem through gfid directly
.TP
\fBro\fR
Mount the filesystem read-only
.TP
\fBenable\-ino32=\fRBOOL
Use 32-bit inodes when mounting to workaround broken applications that don't
support 64-bit inodes
.TP
\fBmem\-accounting
Enable internal memory accounting
.TP
\fBcapability
Enable file capability setting and retrival
.TP
\fBthin-client
Enables thin mount and connects via gfproxyd daemon
.TP
\fBlocaltime-logging
Use local timestamps instead of UTC in mount log entries

.PP
.SS "Advanced options"
.PP
.TP
\fBattribute\-timeout=\fRSECONDS
Set attribute timeout to SECONDS for inodes in fuse kernel module [default: 1]
.TP
\fBentry\-timeout=\fRSECONDS
Set entry timeout to SECONDS in fuse kernel module [default: 1]
.TP
\fBbackground\-qlen=\fRN
Set fuse module's background queue length to N [default: 64]
.TP
\fBgid\-timeout=\fRSECONDS
Set auxiliary group list timeout to SECONDS for fuse translator [default: 0]
.TP
\fBnegative\-timeout=\fRSECONDS
Set negative timeout to SECONDS in fuse kernel module [default: 0]
.TP
\fBvolume\-name=\fRVOLUME-NAME
Volume name to be used for MOUNT-POINT [default: top most volume in
VOLUME-FILE]
.TP
\fBdirect\-io\-mode=\fRBOOL|auto
Specify fuse direct I/O strategy [default: auto]
.TP
\fBcongestion\-threshold=\fRN
Set fuse module's congestion threshold to N [default: 48]
.TP
\fsubdir\-mount=\fRN
Set the subdirectory mount option [default: NULL, ie, no subdirectory mount]
.TP
.TP
\fBbackup\-volfile\-servers=\fRSERVERLIST
Provide list of backup volfile servers in the following format [default: None]
If use IPv6 address in SERVERLIST, it is required to use '[' and ']' to surround IPv6 address.

\fB$ mount \-t glusterfs \-obackup\-volfile\-servers=<server2>:\fR
\fB       <server3>:...:<serverN> <server1>:/<volname> <mount_point>\fR
\fB$ mount \-t glusterfs \-obackup\-volfile\-servers=<[ipv6_addr2]>:\fR
\fB       <[ipv6_addr3]>:...:<[ipv6_addrN]> <ipv6_addr1>:/<volname> <mount_point>\fR

.TP
.TP
\fBbackupvolfile\-server=\fRSERVER
Provide list of backup volfile servers in the following format [default: None]

\fB $ mount \-t glusterfs \-obackupvolfile\-server=<server2>
\fB         <server1>:/<volname> <mount_point>

.TP
.TP
\fBfetch-attempts=\fRN
\fBDeprecated\fR option - placed here for backward compatibility [default: 1]
.TP
.TP
\fBlru-limit=\fRN
Set fuse module's limit for number of inodes kept in LRU list to N [default: 65000]
.TP
.TP
\fBinode-table-size=\fRN
Set the inode hash table size to N - this must be a power of 2 greater than or
equal to 65536 and will be rounded up if not [default: 65536]
.TP
.TP
\fBinvalidate-limit=\fRN
Suspend fuse invalidations implied by 'lru-limit' if the number of outstanding
invalidations I meets the following condition:

If N is an integer: I is equal or bigger than N

If N is a percentage: I is equal or bigger than N percent of lru-limit

[default: 200%]
.TP
.TP
\fBbackground-qlen=\fRN
Set fuse module's background queue length to N [default: 64]
.TP
\fBno\-root\-squash=\fRBOOL
disable root squashing for the trusted client [default: off]
.TP
\fBroot\-squash=\fRBOOL
enable root squashing for the trusted client [default: on]
.TP
\fBuse\-readdirp=\fRBOOL
Use readdirp() mode in fuse kernel module [default: on]
.TP
\fBdump\-fuse=\fRPATH
Dump fuse traffic to PATH
.TP
\fBkernel\-writeback\-cache=\fRBOOL
Enable fuse in-kernel writeback cache [default: off]
.TP
\fBattr\-times\-granularity=\fRNS
Declare supported granularity of file attribute [default: 0]
.TP
\fBauto\-invalidation=\fRBOOL
controls whether fuse-kernel can auto-invalidate attribute, dentry and
page-cache. Disable this only if same files/directories are not
accessed across two different mounts concurrently [default: on]
.PP
.SH FILES
.TP
.I /etc/fstab
A typical GlusterFS entry in /etc/fstab looks like below

\fBserver1:/mirror  /mnt/mirror  glusterfs log-file=/var/log/mirror.log,acl   0  0\fR

.TP
.I /proc/mounts
An example entry of a GlusterFS mountpoint in /proc/mounts looks like below

\fBserver1:/mirror /mnt/glusterfs fuse.glusterfs rw,allow_other,default_permissions,max_read=131072 0 0\fR

.SH SEE ALSO
\fBglusterfs\fR(8), \fBmount\fR(8), \fBgluster\fR(8)

.SH COPYRIGHT
Copyright(c) 2006-2013   Red Hat, Inc.   <http://www.redhat.com>


NEWS
Gluster moves to Gerrit.


contrib/fuse-lib/mount.c
/*
  FUSE: Filesystem in Userspace
  Copyright (C) 2001-2007  Miklos Szeredi <miklos@szeredi.hu>
  Copyright (c) 2010 Gluster, Inc. <http://www.gluster.com>

  This program can be distributed under the terms of the GNU LGPLv2.
  See the file COPYING.LIB.
*/

#include "mount_util.h"
#include "mount-gluster-compat.h"

#ifdef GF_FUSERMOUNT
#define FUSERMOUNT_PROG FUSERMOUNT_DIR "/fusermount-glusterfs"
#else
#define FUSERMOUNT_PROG "fusermount"
#endif
#define FUSE_DEVFD_ENV "_FUSE_DEVFD"

#ifdef __FreeBSD__
#include <sys/types.h>
#include <sys/uio.h>
#include <unistd.h>
#endif /* __FreeBSD__ */

/* FUSE: function is called fuse_kern_unmount() */
void
gf_fuse_unmount(const char *mountpoint, int fd)
{
    int res;
    int pid;

    if (!mountpoint)
        return;

    if (fd != -1) {
        struct pollfd pfd;

        pfd.fd = fd;
        pfd.events = 0;
        res = poll(&pfd, 1, 0);
        /* If file poll returns POLLERR on the device file descriptor,
           then the filesystem is already unmounted */
        if (res == 1 && (pfd.revents & POLLERR))
            return;

        /* Need to close file descriptor, otherwise synchronous umount
           would recurse into filesystem, and deadlock */
        close(fd);
    }

    if (geteuid() == 0) {
        fuse_mnt_umount("fuse", mountpoint, mountpoint, 1);
        return;
    } else {
        GFFUSE_LOGERR("fuse: Effective-uid: %d", geteuid());
    }

    res = umount2(mountpoint, 2);
    if (res == 0)
        return;

    GFFUSE_LOGERR("fuse: failed to unmount %s: %s", mountpoint,
                  strerror(errno));
    pid = fork();
    if (pid == -1)
        return;

    if (pid == 0) {
        const char *argv[] = {FUSERMOUNT_PROG, "-u", "-q", "-z", "--",
                              mountpoint,      NULL};

        execvp(FUSERMOUNT_PROG, (char **)argv);
        GFFUSE_LOGERR("fuse: failed to execute fuserumount: %s",
                      strerror(errno));
        _exit(1);
    }
    waitpid(pid, NULL, 0);
}

/* gluster-specific routines */

/* Unmounting in a daemon that lurks 'till main process exits */
int
gf_fuse_unmount_daemon(const char *mountpoint, int fd)
{
    int ret = -1;
    pid_t pid = -1;

    if (fd == -1)
        return -1;

    int ump[2] = {
        0,
    };

    ret = pipe(ump);
    if (ret == -1) {
        close(fd);
        return -1;
    }

    pid = fork();
    switch (pid) {
        case 0: {
            char c = 0;
            sigset_t sigset;

            close_fds_except(ump, 1);

            setsid();
            (void)chdir("/");
            sigfillset(&sigset);
            sigprocmask(SIG_BLOCK, &sigset, NULL);

            read(ump[0], &c, 1);

            gf_fuse_unmount(mountpoint, fd);
            exit(0);
        }
        case -1:
            close(fd);
            fd = -1;
            ret = -1;
            close(ump[1]);
    }
    close(ump[0]);

    return ret;
}

static char *
escape(char *s)
{
    size_t len = 0;
    char *p = NULL;
    char *q = NULL;
    char *e = NULL;

    for (p = s; *p; p++) {
        if (*p == ',')
            len++;
        len++;
    }

    e = CALLOC(1, len + 1);
    if (!e)
        return NULL;

    for (p = s, q = e; *p; p++, q++) {
        if (*p == ',') {
            *q = '\\';
            q++;
        }
        *q = *p;
    }

    return e;
}

static int
fuse_mount_fusermount(const char *mountpoint, char *fsname, char *mnt_param,
                      int fd)
{
    int pid = -1;
    int res = 0;
    int ret = -1;
    char *fm_mnt_params = NULL;
    char *efsname = NULL;

#ifndef GF_FUSERMOUNT
    GFFUSE_LOGERR(
        "Mounting via helper utility "
        "(unprivileged mounting) is supported "
        "only if glusterfs is compiled with "
        "--enable-fusermount");
    return -1;
#endif

    efsname = escape(fsname);
    if (!efsname) {
        GFFUSE_LOGERR("Out of memory");

        return -1;
    }
    ret = asprintf(&fm_mnt_params, "%s,fsname=%s,nonempty,subtype=glusterfs",
                   mnt_param, efsname);
    FREE(efsname);
    if (ret == -1) {
        GFFUSE_LOGERR("Out of memory");

        goto out;
    }

    /* fork to exec fusermount */
    pid = fork();
    if (pid == -1) {
        GFFUSE_LOGERR("fork() failed: %s", strerror(errno));
        ret = -1;
        goto out;
    }

    if (pid == 0) {
        char env[10];
        const char *argv[32];
        int a = 0;

        argv[a++] = FUSERMOUNT_PROG;
        argv[a++] = "-o";
        argv[a++] = fm_mnt_params;
        argv[a++] = "--";
        argv[a++] = mountpoint;
        argv[a++] = NULL;

        snprintf(env, sizeof(env), "%i", fd);
        setenv(FUSE_DEVFD_ENV, env, 1);
        execvp(FUSERMOUNT_PROG, (char **)argv);
        GFFUSE_LOGERR("failed to exec fusermount: %s", strerror(errno));
        _exit(1);
    }

    ret = waitpid(pid, &res, 0);
    ret = (ret == pid && res == 0) ? 0 : -1;
out:
    FREE(fm_mnt_params);
    return ret;
}

#if defined(__FreeBSD__)
void
build_iovec(struct iovec **iov, int *iovlen, const char *name, void *val,
            size_t len)
{
    int i;
    if (*iovlen < 0)
        return;
    i = *iovlen;

    *iov = realloc(*iov, sizeof **iov * (i + 2));
    if (*iov == NULL) {
        *iovlen = -1;
        return;
    }

    (*iov)[i].iov_base = strdup(name);
    (*iov)[i].iov_len = strlen(name) + 1;

    i++;
    (*iov)[i].iov_base = val;
    if (len == (size_t)-1) {
        if (val != NULL)
            len = strlen(val) + 1;
        else
            len = 0;
    }
    (*iov)[i].iov_len = (int)len;
    *iovlen = ++i;
}

/*
 * This function is needed for compatibility with parameters
 * which used to use the mount_argf() command for the old mount() syscall.
 */
void
build_iovec_argf(struct iovec **iov, int *iovlen, const char *name,
                 const char *fmt, ...)
{
    va_list ap;
    char val[255] = {0};

    va_start(ap, fmt);
    vsnprintf(val, sizeof(val), fmt, ap);
    va_end(ap);
    build_iovec(iov, iovlen, name, strdup(val), (size_t)-1);
}
#endif /* __FreeBSD__ */

struct mount_flags {
    const char *opt;
    mount_flag_t flag;
    int on;
} mount_flags[] = {
    /* We provide best effort cross platform support for mount flags by
     * defining the ones which are commonly used in Unix-like OS-es.
     */
    {"ro", MS_RDONLY, 1},
    {"nosuid", MS_NOSUID, 1},
    {"nodev", MS_NODEV, 1},
    {"noatime", MS_NOATIME, 1},
    {"noexec", MS_NOEXEC, 1},
#ifdef GF_LINUX_HOST_OS
    {"rw", MS_RDONLY, 0},
    {"suid", MS_NOSUID, 0},
    {"dev", MS_NODEV, 0},
    {"exec", MS_NOEXEC, 0},
    {"async", MS_SYNCHRONOUS, 0},
    {"sync", MS_SYNCHRONOUS, 1},
    {"atime", MS_NOATIME, 0},
    {"dirsync", MS_DIRSYNC, 1},
#endif
    {NULL, 0, 0}};

static int
mount_param_to_flag(char *mnt_param, mount_flag_t *mntflags,
                    char **mnt_param_new)
{
    gf_boolean_t found = _gf_false;
    struct mount_flags *flag = NULL;
    char *param_tok = NULL;
    token_iter_t tit = {
        0,
    };
    gf_boolean_t iter_end = _gf_false;

    /* Allocate a buffer that will hold the mount parameters remaining
     * after the ones corresponding to mount flags are processed and
     * removed.The length of the original params are a good upper bound
     * of the size needed.
     */
    *mnt_param_new = strdup(mnt_param);
    if (!*mnt_param_new)
        return -1;

    for (param_tok = token_iter_init(*mnt_param_new, ',', &tit);;) {
        iter_end = next_token(&param_tok, &tit);

        found = _gf_false;
        for (flag = mount_flags; flag->opt; flag++) {
            /* Compare the mount flag name to the param
             * name at hand.
             */
            if (strcmp(flag->opt, param_tok) == 0) {
                /* If there is a match, adjust mntflags
                 * accordingly and break.
                 */
                if (flag->on) {
                    *mntflags |= flag->flag;
                } else {
                    *mntflags &= ~flag->flag;
                }
                found = _gf_true;
                break;
            }
        }
        /* Exclude flag names from new parameter list. */
        if (found)
            drop_token(param_tok, &tit);

        if (iter_end)
            break;
    }

    return 0;
}

static int
fuse_mount_sys(const char *mountpoint, char *fsname, char *mnt_param, int fd)
{
    int ret = -1;
    unsigned mounted = 0;
    char *mnt_param_mnt = NULL;
    char *fstype = "fuse.glusterfs";
    char *source = fsname;
    mount_flag_t mountflags = 0;
    char *mnt_param_new = NULL;

    ret = mount_param_to_flag(mnt_param, &mountflags, &mnt_param_new);
    if (ret == 0)
        ret = asprintf(&mnt_param_mnt,
                       "%s,fd=%i,rootmode=%o,user_id=%i,group_id=%i",
                       mnt_param_new, fd, S_IFDIR, getuid(), getgid());
    if (ret == -1) {
        GFFUSE_LOGERR("Out of memory");

        goto out;
    }

#ifdef __FreeBSD__
    struct iovec *iov = NULL;
    int iovlen = 0;
    char fdstr[15];
    sprintf(fdstr, "%d", fd);

    build_iovec(&iov, &iovlen, "fstype", "fusefs", -1);
    build_iovec(&iov, &iovlen, "subtype", "glusterfs", -1);
    build_iovec(&iov, &iovlen, "fspath", __DECONST(void *, mountpoint), -1);
    build_iovec(&iov, &iovlen, "from", "/dev/fuse", -1);
    build_iovec(&iov, &iovlen, "volname", source, -1);
    build_iovec(&iov, &iovlen, "fd", fdstr, -1);
    build_iovec(&iov, &iovlen, "allow_other", NULL, -1);
    ret = nmount(iov, iovlen, mountflags);
#else
    ret = mount(source, mountpoint, fstype, mountflags, mnt_param_mnt);
#endif /* __FreeBSD__ */
#ifdef GF_LINUX_HOST_OS
    if (ret == -1 && errno == ENODEV) {
        /* fs subtype support was added by 79c0b2df aka
           v2.6.21-3159-g79c0b2d. Probably we have an
           older kernel ... */
        fstype = "fuse";
        ret = asprintf(&source, "glusterfs#%s", fsname);
        if (ret == -1) {
            GFFUSE_LOGERR("Out of memory");

            goto out;
        }
        ret = mount(source, mountpoint, fstype, mountflags, mnt_param_mnt);
    }
#endif /* GF_LINUX_HOST_OS */
    if (ret == -1)
        goto out;
    else
        mounted = 1;

#ifdef GF_LINUX_HOST_OS
    if (geteuid() == 0) {
        char *newmnt = fuse_mnt_resolve_path("fuse", mountpoint);
        char *mnt_param_mtab = NULL;

        if (!newmnt) {
            ret = -1;

            goto out;
        }

        ret = asprintf(&mnt_param_mtab, "%s%s",
                       mountflags & MS_RDONLY ? "ro," : "", mnt_param_new);
        if (ret == -1)
            GFFUSE_LOGERR("Out of memory");
        else {
            ret = fuse_mnt_add_mount("fuse", source, newmnt, fstype,
                                     mnt_param_mtab);
            FREE(mnt_param_mtab);
        }

        FREE(newmnt);
        if (ret == -1) {
            GFFUSE_LOGERR("failed to add mtab entry");

            goto out;
        }
    }
#endif /* GF_LINUX_HOST_OS */

out:
    if (ret == -1) {
        GFFUSE_LOGERR("ret = -1\n");
        if (mounted)
            umount2(mountpoint, 2); /* lazy umount */
    }
    FREE(mnt_param_mnt);
    FREE(mnt_param_new);
    if (source != fsname)
        FREE(source);

    return ret;
}

int
gf_fuse_mount(const char *mountpoint, char *fsname, char *mnt_param,
              pid_t *mnt_pid, int status_fd)
{
    int fd = -1;
    pid_t pid = -1;
    int ret = -1;

    fd = open("/dev/fuse", O_RDWR);
    if (fd == -1) {
        GFFUSE_LOGERR("cannot open /dev/fuse (%s)", strerror(errno));
        return -1;
    }

    /* start mount agent */
    pid = fork();
    switch (pid) {
        case 0:
            /* hello it's mount agent */
            if (!mnt_pid) {
                /* daemonize mount agent, caller is
                 * not interested in waiting for it
                 */
                pid = fork();
                if (pid)
                    exit(pid == -1 ? 1 : 0);
            }

            ret = fuse_mount_sys(mountpoint, fsname, mnt_param, fd);
            if (ret == -1) {
                gf_log("glusterfs-fuse", GF_LOG_INFO,
                       "direct mount failed (%s) errno %d", strerror(errno),
                       errno);

                if (errno == EPERM) {
                    gf_log("glusterfs-fuse", GF_LOG_INFO,
                           "retry to mount via fusermount");

                    ret = fuse_mount_fusermount(mountpoint, fsname, mnt_param,
                                                fd);
                }
            }

            if (ret == -1)
                GFFUSE_LOGERR("mount of %s to %s (%s) failed", fsname,
                              mountpoint, mnt_param);

            if (status_fd >= 0)
                (void)write(status_fd, &ret, sizeof(ret));
            exit(!!ret);
            /* bye mount agent */
        case -1:
            close(fd);
            fd = -1;
    }

    if (mnt_pid)
        *mnt_pid = pid;

    return fd;
}
